packages/core/src/context/layer2.ts

```diff
@@ -284,41 +282,38 @@ export class L2Chain {
       }
       return a.logIndex - b.logIndex
     })
-    leaves.push(...pendingDeposits.map(deposit => Fp.from(deposit.note)))
-    consumedBytes += commits.length
-    aggregatedFee = aggregatedFee.add(
-      pendingDeposits.reduce((prev, item) => prev.add(item.fee), Fp.zero),
-    )
-    const includedIndexes = {}
-    const validLeaves = [] as Fp[]
+    // 3. validation
+    const validCommits = [] as MassDepositSql[]
+    const validDeposits = [] as DepositSql[]
     for (const commit of commits) {
       const deposits = pendingDeposits.filter(deposit => {
         return deposit.queuedAt === commit.index
       })
+      // If found missing deposit or no deposit in commits,
+      // stop iteration
+      if (deposits.length === 0) {
+        break
+      }
       const { merged, fee } = mergeDeposits(deposits)
       if (
         merged.toString() !== commit.merged ||
         !Fp.from(fee.toString()).eq(Fp.from(commit.fee))
       ) {
-        // eslint-disable-next-line no-continue
-        continue
+        break
       }
-      validLeaves.push(...deposits.map(deposit => Fp.from(deposit.note)))
-      includedIndexes[commit.index] = true
+      validCommits.push(commit)
+      validDeposits.push(...deposits)
     }
     return {
-      massDeposits: commits
-        .filter(commit => includedIndexes[commit.index])
-        .map(commit => ({
-          merged: Bytes32.from(commit.merged),
-          fee: Uint256.from(commit.fee),
-        })),
-      leaves: validLeaves,
-      totalFee: commits.reduce((acc, commit) => {
-        if (!includedIndexes[commit.index]) return acc
+      massDeposits: validCommits.map(commit => ({
+        merged: Bytes32.from(commit.merged),
+        fee: Uint256.from(commit.fee),
+      })),
+      leaves: validDeposits.map(deposit => Fp.from(deposit.note)),
+      totalFee: validCommits.reduce((acc, commit) => {
         return acc.add(Fp.from(commit.fee))
       }, Fp.zero),
-      calldataSize: consumedBytes,
+      calldataSize: validCommits.length,
```

Comment:
I think we're missing consumedBytes logic here at `block-generator.ts` and here. IMO, I think we can remove this `calldataSize` property here and improve the logic at `block-generator.ts`